### ISSUE TITLE
V1.5.0. RC2 Nightlys    VAR need steps in exact  us not in %  for a Counter inc/dec

### ISSUE THREAD

**Issue by hrenz** [FrSkyRC/ETHOS-Feedback-Community#3427]
At VAR or ExtraTrim I can count (inc/dec) in Steps in %

At  trim with extra fein , fine,  medium,  coare,   the  steps are in 1, 2, 4, 8 us

But I need exakt steps in us  10us, 25us ,50us, 100us and so but not in %

Because  1% are 5,12 us  and    5% are 25,6us   ( 512us/100% = 5,12us)

That's not accurate

To query logical switches I have the same problem there with %

What to do in VAR or in ExtraTrim

**Comment by mawzthefinn** [comment]
Not possible given the architecture. Mixers calculate based on the 2048 step range (mapped to percent).

Mixer outputs are then mapped to the endpoints sent to the receiver, this is displayed as uS, but the mapping to uS in the Outputs is variable depending on the selected endpoints and PWM center for each channel. You can readily setup a model so the % to uS mapping differs on each side of center (and even over the range if using an Output curve).

If you set your channel endpoints so they are +/- 500uS instead of +/-512uS default (ie -100 is at 1000uS, 100 at 2000uS), you will get the mapping you want, but lose the ability to define endpoints based on throw.

**Comment by JimB40** [comment]
Did some testing on 1.5 RC and to also to my surprise Var now has constant % unit and allows only Sources that have % unit
So what exactly Vars evolved into? Tool to manipulate sources used in Mixer?

My standard example is calculating pack's cell number
Formula is simple you take RxBt or VFAS and divide by pack cell voltage (4.2-4.35 depending on lipo type)

So now above mentioned example rendered impossible.
1. Vars are not accepting sources other than % (ie. Telemetry sensor)
2. Calculated sensor does not allow to perform math operations like Vars (ie. Add/subtract/multiply... etc)

AFAIU Vars now in 1.5 RC Ethos are not variables (like GVAR in OpenTX) when you can set any numeric value and manipulate it.
Of course one of cases can be percent
To give example:
If you set variable value to number=2010 it is stored internally as 2010. But if you then perform action 'Percent=100' it becomes 20.10 internally.

ETHOS logic can also, in case of sources used in mixer, forcefully set variable unit to % and variable min,max limits to -100,100
But ditching in Vars values other than % seems to me as step backward.

**Comment by bsongis-frsky** [comment]
@JimB40 I would rather use a Calculated Sensor for your use case, no?

**Comment by bsongis-frsky** [comment]
A screenshot to be sure the proposed solution fits to your need ...

![image](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/50834778/8be407b8-b8d7-458e-a81d-5000b3a8225c)

**Comment by bsongis-frsky** [comment]
Any other use case where a Var would need something else than %
Scratching my head to see something interesting to work on :)

**Comment by JimB40** [comment]
Yes if LiPo Cell1 is kind of GVAR not telemetry sensor.
AFAIR D16 in not sending every cell voltage  just total VFAS

Okay here is real life example:
Rather than having 30 models in radio we have few, named after rclink/firmware ie D16 BF, D16 INAV, ELRS BF, CFRS BF etc
Setup in radio is common so all it takes with new aircraft is to configure properly its FC firmware.

So one of common tasks is to set lipo low and critical alarms.
Having different aircrafts with lipo from 1 to 6, here is what I did with first 1.5 versions:

1. Set var called "Cell No" so its value changed from 1 to 6 based on FS pressed (LED gives me confirmation how many cells have lipo I've connected"
2. Set var called "Cell Avg V" = VFAS (telemetry sensor) / "Cell No" (var changed by FS)
3. Set up logic switch "Batt Low" A<X "Cell Avg V" < 3.6
4. Set up special function to play warning with "Batt Low" as Acive condition.

So no matter what aircraft was connected low battery warning was always working properly, only thing to remember was to press corresponding FS.

How to set it up with RC2? Don't have bloody idea even tried few ways. :)

**Comment by bsongis-frsky** [comment]
LiPo Cell1 is a Telemetry sensor in my screenshot.

![image](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/50834778/5fa186c2-0f01-4f02-878a-06cdef906e81)

**Comment by hrenz** [comment]
My problem with VAR and Trim
is that I cannot use them to realise an exact counter and then query it for an exact counter value.

with exact adjustable steps as inc/dec 

Whether there are pure numbers internally 
or internally as us with a range of 2048
would not matter to me.
They must not be changed depending on the range.
The range is used as real limit value monitoring

**Comment by bsongis-frsky** [comment]
We are mixing 2 different discussions here, it becomes hard to follow.

I have just added (for RC4) the possibility to have a Var in Custom sensor calculations

![image](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/50834778/c59771b9-6357-4072-ba1f-092522728e70)

**Comment by bsongis-frsky** [comment]
@JimB40 I have understood your need now. It makes sense. I think you don't need 1.5.0, it should be already possible to do with 1.4.X

- Set var called "Cell No" so its value changed from 1 to 6 based on FS pressed (LED gives me nice confirmation pack with how many cells I'm going to fly"

![image](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/50834778/4f977e85-2417-4823-bdf4-2f4f4cd65698)

- Set var called "Cell Avg V" = VFAS (telemetry sensor) / "Cell No" (var changed by FS)

![image](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/50834778/ee38c275-f30e-43fa-9390-8557558e921a)

- Set up logic switch "Batt Low" A<X "Cell Avg V" < 3.6

![image](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/50834778/aad10d58-b420-460b-9cbd-025bef5d2f0a)

- Set up special function to play warning with "Batt Low" as Acive condition.

No problem here

**Comment by bsongis-frsky** [comment]
To could also use a Divide operator to avoid the loss of precision in 33.3 and 16.6, the Var should be adapted with 60% / 50% / 40% / 30% / 20% / 10% (and with a divide operator, yes the unit could make sense)

I wonder if I could merge the "Custom sensor" with the "Var" in a future release to make something even more powerful:
- Custom sensor would benefit from the Actions implemented in Vars
- Vars would benefit from the Unit / Precision implemented in Custom Sensors

And if done with the proposal from #3203 you could only need one Var to do the job:

```
Var "Calculated Cell"
=============
Source = VFAS
Unit = V
Precision = 2
Actions
   FS1 ... Divide by ... 1
   FS2 ... Divide by ... 2
   FS3 ... Divide by ... 3
   FS4 ... Divide by ... 4
   FS5 ... Divide by ... 5
   FS6 ... Divide by ... 6
```

**Comment by JimB40** [comment]
Thanks Bernard for use case. 
As they say "Different roads lead to Rome" :)
No need to over-complicate OS as testing, edge cases and manual creation becomes nightmare then.

**Comment by bsongis-frsky** [comment]
The idea of merging "Calculated Sensor" with "Var" seems to me a good idea. If it had existed since the beginning you would have found your road without opening an issue ;) 

The Var would benefit from "Write Logs" too, which seems not bad.

Moreover, and to return to @hrenz original issue I think it would give him a solution. I would like to be sure of the purpose of his counter, it could be misused here. But here is the beginning of a proposal:

```
Var "HRenz Counter"
=============
Source = VFAS
Unit = --- (None)
Precision = 0
Actions
   SA(up) ...      Add and store ... 1
   SA(down) ... Add and store ... -1
```

**Comment by agibson2** [comment]
I was trying to find a way to calculate descent rate for testing my DLG elevator trim and camber settings to find minimum sink.  The idea was to trigger at a certain height by switch (save the current altitude in a Var) and then after 30 seconds take the current altitude and the difference divided by 30 seconds would be the descent rate.  The Var feature you are talking about possibly adding wouldn't allow storing a value would it?

I guess if we could get access to adjust the vspeed averaging for how long it averages the data, that would work without any special modifications.

**Comment by hrenz** [comment]
we only need real counters 
with inc, dec, in adjustable steps and range 
which can be set with preset, reset
and can be queried exactly with log switches
Ideal in us and a counting range from approx. -1000 to+1000
The application and the requirement are quite simple

**Comment by bsongis-frsky** [comment]
To fix this issue:
- Allow a range [-1024:+1024]
- Allow decimals = 0
- Set unit = RAW

Merged into #3821
